plugin/trino-kudu/src/test/java/io/trino/plugin/kudu/TestingKuduServer.java

```diff
@@ -15,65 +15,184 @@
+     * In order to workaround this, we create volume mounts into each container and startup each container and have them watch the volume mounts
+     * When the container starts up, we get the mapped port and write into the volume mount a special "mapped_port" file saying what the mapped port is
+     * Then we can configure --rpc_advertised_addresses and startup kudu
+     * Finally, since we never actually checked the containers started up correctly, we manually fire the HostPortWaitStrategy
+     */
     public TestingKuduServer()
     {
         Network network = Network.newNetwork();
         ImmutableList.Builder<GenericContainer<?>> tServersBuilder = ImmutableList.builder();
-        this.master = new GenericContainer<>("apache/kudu:1.10.0")
+
+        String hostIP = getHostIPAddress();
+        createHostTmpDirectoriesForVolumeMounts();
+
+        String masterContainerAlias = "kudu-master";
+        this.master = new GenericContainer<>("apache/kudu:1.15.0")
+                .withFileSystemBind(masterTmpDirectory, "/testcontainers")
+                .withCopyFileToContainer(forClasspathResource("wait_until_port_bound.sh", 0777), "/wait_until_port_bound.sh")
                 .withExposedPorts(KUDU_MASTER_PORT)
-                .withCommand("master")
+                .withCommand("/wait_until_port_bound.sh", "master")
+                .withEnv("KUDU_MASTERS", "kudu-master:7051")
+                // Note `rpc_advertised_addresses` does not have a port configured, it will be configured after container startup in `/wait_until_port_bound.sh`
+                .withEnv("MASTER_TEMPLATE_ARGS", format("--fs_wal_dir=/var/lib/kudu/master --logtostderr --use_hybrid_clock=false --rpc_bind_addresses=%s:%s --rpc_advertised_addresses=%s:", masterContainerAlias, KUDU_MASTER_PORT, hostIP))
                 .withNetwork(network)
-                .withNetworkAliases("kudu-master");
+                .withNetworkAliases(masterContainerAlias)
+                // Do not wait, the host port will not be listening until after the container starts up and gets the mapped host port
+                .waitingFor(new NoWaitStrategy());
 
         for (int instance = 0; instance < NUMBER_OF_REPLICA; instance++) {
             String instanceName = "kudu-tserver-" + instance;
-            GenericContainer<?> tableServer = new GenericContainer<>("apache/kudu:1.10.0")
+            GenericContainer<?> tableServer = new GenericContainer<>("apache/kudu:1.15.0")
+                    .withFileSystemBind(tabletTmpDirectories.get(instance), "/testcontainers")
+                    .withCopyFileToContainer(forClasspathResource("wait_until_port_bound.sh", 0777), "/wait_until_port_bound.sh")
                     .withExposedPorts(KUDU_TSERVER_PORT)
-                    .withCommand("tserver")
+                    .withCommand("/wait_until_port_bound.sh", "tserver")
```

Comment:
Instead of using a custom script to bind the address, can we use `ToxiproxyContainer` from TestContainers, specify a proxy and then use the mapped proxy port here.

It could be something like this 
```
        for (int instance = 0; instance < NUMBER_OF_REPLICA; instance++) {
            String instanceName = "kudu-tserver-" + instance;
            ContainerProxy proxy = toxiProxy.getProxy(instanceName, KUDU_TSERVER_PORT);
            GenericContainer<?> tableServer = new GenericContainer<>("apache/kudu:1.15.0")
                    .withExposedPorts(KUDU_TSERVER_PORT)
                    .withEnv("KUDU_MASTERS", "kudu-master:" + KUDU_MASTER_PORT)
                    .withEnv("TSERVER_ARGS", format("--fs_wal_dir=/var/lib/kudu/tserver --logtostderr --use_hybrid_clock=false --rpc_bind_addresses=%s:%s --rpc_advertised_addresses=%s:%s", instanceName, KUDU_TSERVER_PORT, hostIP, proxy.getProxyPort()))
                    .withNetwork(network)
                    .withNetworkAliases(instanceName)
                    .waitingFor(new HostPortWaitStrategy())
                    .withCommand("tserver")
                    .dependsOn(master);

            tServersBuilder.add(tableServer);
        }
```

I guess we don't have to bind file in thos case. WDYT ?

Reply:
Thanks for fixing this.

Reply:
That would definitely be easier than messing with bind mounts! Let me see if I can get it working.

Reply:
Made things much cleaner, thanks for the suggestion!